operator: per-container resource requests, v1beta API and CRD webhook

avalluri opened the pull request:
The changes in brief:
- Split the deployment API code such that, could be reused for multiple API version.
- Provided per-container resources as part of v1beta1 API.
- Conversion webhook for migrating existing v1alpha1 API to v1beta1. [Not yet supported deploying using OLM]

pohly reviewed (commented):
> Conversion webhook for migrating existing v1alpha1 API to v1beta1. [Not yet supported deploying using OLM]

What's the situation when deploying with OLM? Only v1beta1 API available or operator not usable?

avalluri commented:
> What's the situation when deploying with OLM? Only v1beta1 API available or operator not usable?

Yes, It accepts v1beta1 CRs. For v1alpha1 CR, the default conversion works, i.e, only the `apiVersion` field. This means any v1alpha1 deployment that was deployed with default values will automatically be converted to v1beta1. Only v1alpha1 deployments which explicitly use `{node,controller}Resources` fields will not work.

pohly commented:
> Yes, It accepts v1beta1 CRs. For v1alpha1 CR, the default conversion works, i.e, only the apiVersion field. This means any v1alpha1 deployment that was deployed with default values will automatically be converted to v1beta1. Only v1alpha1 deployments which explicitly use {node,controller}Resources fields will not work as they are deprecated in the new API.

This needs to be documented in `install.md`.

avalluri commented:
> > Yes, It accepts v1beta1 CRs. For v1alpha1 CR, the default conversion works, i.e, only the apiVersion field. This means any v1alpha1 deployment that was deployed with default values will automatically be converted to v1beta1. Only v1alpha1 deployments which explicitly use {node,controller}Resources fields will not work as they are deprecated in the new API.
> 
> This needs to be documented in `install.md`.

Current install steps did not cover installing the operator using OLM. I added this to the install steps and also added a note about the lack of webhook support.

pohly commented on docs/install.md:
```diff
@@ -1302,6 +1314,45 @@ deployment is in the `Failed` state, then one can look into the event(s) using
+## PMEM-CSI operator API conversion webhook
+
+The operator relese v0.9.0 deprecates the `v1alpha1` [deployment API](#pmem-csi-deployment-crd) infavour of
+`v1beta1`. To support auto-migrating of deprecated(v1alpha1) API versions, The
```
Please explain what "auto-migrating" does. Be specific (i.e. provide examples) to illustrate it. This is also the right place to describe what happens without the webhook. We have to assume that the reader is not familiar with CRD concepts.

avalluri replied:
Added more explanation about CRD conversion and added suitable examples to the docuentation.

pohly commented on docs/install.md:
```diff
@@ -1302,6 +1314,45 @@ deployment is in the `Failed` state, then one can look into the event(s) using
+service named `pmem-cis-operator-service` in the operator's namespace.
+
+To secure the communication between the Kubernetes APIserver and the conversion
+webhook, the server must be configured with apporopriate private key and a TLS
```
Please run a spell-checker ("apporopriate").

avalluri replied:
Fixed.

pohly commented on docs/install.md:
```diff
@@ -1302,6 +1314,45 @@ deployment is in the `Failed` state, then one can look into the event(s) using
+
+**NOTE**: Conversion webhook is not yet supported by the operator installed using the
+[Operator Lifecycle Manager](https://operatorhub.io/operator/pmem-csi-operator).
+If the operator is deployed with no conversion webhook then it might
```
"might not" or "will not"? "Might" sounds like it's uncertain.

avalluri replied:
Rephrased to "will not".

pohly commented on operator/operator.make:
```diff
@@ -1,15 +1,15 @@
 	tmpdir=`mktemp -d` && \
 	trap 'set -x; rm -rf $$tmpdir' EXIT && \
-	git clone --branch 1.0.0+fixes https://github.com/avalluri/operator-sdk.git $$tmpdir && \
+	git clone --branch $(OPERATOR_SDK_VERSION)+fixes https://github.com/avalluri/operator-sdk.git $$tmpdir && \
```
Do we still need additional fixes? Please add links to pending PRs for all of them so that we can check progress.

avalluri replied:
The pending PR link(s) are available at few lines above in make target.

pohly replied:
Looking at https://github.com/operator-framework/operator-sdk/pull/3786 I see that it was closed as "open for a while without interation". It looks like there was a test failure that didn't get addressed for two months?

Can you do what was suggested in https://github.com/operator-framework/operator-sdk/pull/3786#issuecomment-719659176 and create an issue? Then if you have a working fix, create a PR.

pohly replied:
Please work with upstream and add a link to that work (probably the new issue) here.

avalluri replied:
Reported the [issue](https://github.com/operator-framework/operator-sdk/issues/4293).

pohly commented on pkg/apis/pmemcsi/v1beta1/deployment_conversion.go:
```diff
@@ -0,0 +1,26 @@
+var _ conversion.Hub = &Deployment{}
+
+// Hub marks this(v1beta1) type as a conversion hub.
+func (*Deployment) Hub() {}
```
Explaining "hub" as "conversion hub" doesn't explain what a "hub" is (circular definition). Please explain what it does.

avalluri replied:
This is copied from the generated code by controller-runtime/operator-sdk. Added a short description:
`// Conversion hub is the version that all other CRD versions get converted into.`

pohly commented on test/e2e/operator/deployment_api.go:
```diff
@@ -775,6 +776,70 @@ var _ = deploy.DescribeForSome("API", func(d *deploy.Deployment) bool {
+
+			validateDriver(deployment, true)
+		})
+	})
```
According to the conversion functions above, conversion from beta to alpha is supported. This is not tested here, is it?

avalluri replied:
Though the conversion code is available, the operator does not reconcile `v1alpha1` CRs, and thus no test to cover that.

avalluri replied:
Extended the tests to make sure that the `v1beta1` ->``v1alpha1` conversion works:
 - create `v1alpha1` object
 - retrieve back the `v1alpha1` revision and make sure that both the `spec` fields match.
 - retrieve the `v1beta1` revision for the same object and check that the `status` matches with the `v1alpah1` status.

pohly replied:
"conversion with explicit values" fails: https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/blue/organizations/jenkins/pmem-csi/detail/PR-826/15/tests

pohly replied:
> Extended the tests to make sure that the v1beta1 ->``v1alpha1` conversion works:

This isn't really converting something that was created as v1beta1, is it?

avalluri replied:
> > Extended the tests to make sure that the v1beta1 ->``v1alpha1` conversion works:
> 
> This isn't really converting something that was created as v1beta1, is it?

The test creates a `v1alpha1` object, which actually stored as the `v1beta1` by the API server after the conversion. Then the test retrieves back both `v1alpah1` and `v1beta1` versions for the same CR, and checks if both have expected fields.

pohly commented:
The way I understand, the current design is:
- operator only supports v1beta1 API
- v1alpha must be converted to v1beta either:
  - on-the-fly via webhook or
  - permanently manually by the user and/or admin

Have you considered supporting both v1alpha1 and and v1beta1 in the operator? Is that possible?

avalluri commented:
> Have you considered supporting both v1alpha1 and and v1beta1 in the operator? Is that possible?

I did not consider supporting both(all) versions. It makes the reconciling code bit complicated. Do you see that we should support both API versions without conversion?

pohly commented on docs/install.md:
```diff
@@ -1215,8 +1223,12 @@ The current API for PMEM-CSI `Deployment` resources is:
 | deviceMode | string | Device management mode to use. Supports one of `lvm` or `direct` | `lvm`
-| controllerResources | [ResourceRequirements](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#resourcerequirements-v1-core) | Describes the compute resource requirements for controller pod |
-| nodeResources | [ResourceRequirements](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#resourcerequirements-v1-core) | Describes the compute resource requirements for the pods running on node(s) |
+| controllerResources | [ResourceRequirements](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#resourcerequirements-v1-core) | Describes the compute resource requirements for controller pod [**_Deprectead_**] |
```
What does "deprecated" mean here? Isn't `controllerResources` gone in v1beta1?

We probably need a more nuanced way of describing the API:
- for `apiVersion`, list v1beta1 and v1alpha1 as valid values
- add a general remark outside of the table that fields are valid in all API revisions unless noted otherwise
- for fields that only exist in some revision(s), list those revisions in the description

avalluri replied:
> What does "deprecated" mean here? Isn't `controllerResources` gone in v1beta1?
> 
Yes, `v1beta1` uses `controllerDriverResources` and `provisionerResources` in place of `controllerResources`.

> We probably need a more nuanced way of describing the API:
> 
> * for `apiVersion`, list v1beta1 and v1alpha1 as valid values
> * add a general remark outside of the table that fields are valid in all API revisions unless noted otherwise
> * for fields that only exist in some revision(s), list those revisions in the description

Made changes as suggested.

pohly commented on docs/install.md:
```diff
@@ -1302,6 +1314,140 @@ deployment is in the `Failed` state, then one can look into the event(s) using
+version. This conversion is done by the Kubernetes API server by using the
+conversion strategy chosen in the [deployment CRD sepcification](#pmem-csi-deployment-crd).
+If no conversion strategy is chosen by the CRD, by default the API server
+converts the custom resource `apiVersion` to the latest supported CRD version.
```
Converts when? When a client asks for v1beta1 (i.e. on demand)?

avalluri replied:
> Converts when? When a client asks for v1beta1 (i.e. on demand)?

As pet the [webhook conversion](https://v1-15.docs.kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#webhook-conversion) documentation:

```
The API server also supports webhook conversions that call an external service in case a conversion is required. For example when:

- custom resource is requested in a different version than stored version.
- Watch is created in one version but the changed object is stored in another version.
- custom resource PUT request is in a different version than storage version.
```

pohly commented on docs/install.md:
````diff
@@ -1302,6 +1314,140 @@ deployment is in the `Failed` state, then one can look into the event(s) using
+$ kubectl get deployments.pmem-csi.intel.com pmem-csi.intel.com -o jsonpath='{.apiVersion}{"\n"}{.spec}{"\n"}'
+pmem-csi.intel.com/v1beta1
+{"deviceMode":"lvm"}
+```
````
It's worth showing here what happens when a client asks for a v1alpha1 version of that same object.

avalluri replied:
> It's worth showing here what happens when a client asks for a v1alpha1 version of that same object.

It supposed to return the same object by setting `apiVersion` to `v1alpha1`

pohly replied:
But with `controllerResources` and `nodeResources` set, so not quite the same. This illustrates what we need to mention earlier about objects being converted on demand instead of being rewritten in the etcd store.

pohly replied:
Long-term that becomes relevant for admins: if we ever *remove* the v1alpha1 API, they need to ensure that none of the stored CRs use that version or they'll become unusable.

avalluri replied:
> But with `controllerResources` and `nodeResources` set, so not quite the same. This illustrates what we need to mention earlier about objects being converted on demand instead of being rewritten in the etcd store.

Only the latest(`v1beta1`)  CR fields get stored and when the client asks for a different revision, then the conversion happens from the stored revision. In this case where no conversion webhook, the `nodeResource`, `controllerResources` fields are unknown to `v1beta1` get lost and hence no change if we ask for `v1beta1` or `v1alpha1`.

pohly commented on pkg/apis/pmemcsi/base/defaults.go:
```diff
@@ -0,0 +1,85 @@
+	// recommendations reported by VirtualPodAutoscaler(LowerBound -> Requests and UpperBound -> Limits)
+
+	// DefaultControllerResourceRequestCPU default CPU resource request used for controller driver container
+	DefaultControllerResourceRequestCPU = "12m" // MilliSeconds
```
Please document under DEVELOPMENT.md how these defaults were obtained. I already added instructions for running Vertical Pod Autoscaler; what's missing is a description of the workload and how to run it.

You said that you used a script. In that case we'll need the script so that others can reproduce your experiment. It might be simpler to use the late binding stress test.

avalluri replied:
Running the stress test is a good enough example, but anyway I also added the alternate to run the script.

pohly reviewed (changes requested):
I'm still mostly looking at documentation to understand what I should expect from the code...

The new documentation contains quite a few typos. Can you run a spell checker and fix those? 

I'll probably then propose some language changes in a PR (will be faster than leaving comments).

avalluri commented:
Seems like the fedora-31 downlink is broken. Is it time to move to the latest release?
```
[2020-12-03T11:09:39.316Z] Downloading https://download.fedoraproject.org/pub/fedora/linux/releases/31/Cloud/x86_64/images/Fedora-Cloud-Base-31-1.9.x86_64.raw.xz image
[2020-12-03T11:09:39.316Z]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[2020-12-03T11:09:39.316Z]                                  Dload  Upload   Total   Spent    Left  Speed
[2020-12-03T11:09:39.673Z] 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
[2020-12-03T11:09:39.673Z] curl: (22) The requested URL returned error: 404 
[2020-12-03T11:09:39.673Z]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[2020-12-03T11:09:39.673Z]                                  Dload  Upload   Total   Spent    Left  Speed
[2020-12-03T11:09:40.032Z] 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
[2020-12-03T11:09:40.032Z] curl: (22) The requested URL returned error: 404 
[2020-12-03T11:09:40.032Z]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[2020-12-03T11:09:40.032Z]                                  Dload  Upload   Total   Spent    Left  Speed
[2020-12-03T11:09:40.390Z] 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
[2020-12-03T11:09:40.390Z] curl: (22) The requested URL returned error: 404 
[2020-12-03T11:09:40.390Z]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[2020-12-03T11:09:40.390Z]                                  Dload  Upload   Total   Spent    Left  Speed
[2020-12-03T11:09:40.746Z] 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
[2020-12-03T11:09:40.746Z] curl: (22) The requested URL returned error: 404 
[2020-12-03T11:09:40.746Z]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[2020-12-03T11:09:40.746Z]                                  Dload  Upload   Total   Spent    Left  Speed
[2020-12-03T11:09:41.104Z] 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
[2020-12-03T11:09:41.104Z] curl: (22) The requested URL returned error: 404 
[2020-12-03T11:09:41.104Z]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[2020-12-03T11:09:41.104Z]                                  Dload  Upload   Total   Spent    Left  Speed
[2020-12-03T11:09:41.461Z] 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
[2020-12-03T11:09:41.461Z] curl: (22) The requested URL returned error: 404 
[2020-12-03T11:09:41.461Z] ERROR: Download failed repeatedly, giving up.
```

pohly commented on hack/stress-driver.sh:
```diff
@@ -0,0 +1,27 @@
+    sed -e "s;\(.*name:\)\(.*\);\1\2-$j;g" < deploy/common/pmem-pvc.yaml | kubectl create -f -
+  done
+  echo "Deleting all pvc..."
+  while [ "$(kubectl get pvc --no-headers | wc -l)" -ne "0" ]; do
```
Is this supposed to wait until all volumes are deleted? Then you need to check for `pv` here, not `pvc`. First a PVC gets deleted, then the PV.

pohly replied:
Ping...

avalluri replied:
Yes, I agree. Will change the check.

avalluri replied:
Done.

pohly commented on docs/install.md:
```diff
@@ -1302,6 +1320,144 @@ deployment is in the `Failed` state, then one can look into the event(s) using
+If the operator is deployed with no conversion webhook then it will not
+support the conversion of deprecated API fields as described above.
+In that case, the user has to manually migrate their existing driver deployments
+to use the latest operator API.
```
Huh? I thought the apiserver will convert internally when no webhook is configured. So why does a user "have to manually migrate their existing driver deployments"?

pohly replied:
Hmm, "in that case" is not clear. It doesn't refer to "deployed with no conversion webhook" (my initial understanding) but rather "conversion of deprecated API fields". Will fix...

pohly commented:
Please remove the "Fedora 33" commit.

Tip: if you have to carry the same commit in two branches, then base the larger branch on top of the smaller one. Then if the smaller branch gets merged, the larger one can also be merged without rebasing because it is based on a commit that is in the target branch.

As it stands now, we would have two different "test cluster: Move to fedora-33 cloud image" commits in the "devel" branch if we merged this PR.

pohly commented on test/e2e/operator/deployment_api.go:
```diff
@@ -690,7 +722,7 @@ var _ = deploy.DescribeForSome("API", func(d *deploy.Deployment) bool {
 				"metrics service": func(dep *api.Deployment) apiruntime.Object {
 					obj := &corev1.Service{}
-					key := runtime.ObjectKey{Name: dep.MetricsServiceName(), Namespace: d.Namespace}
+					key := runtime.ObjectKey{Name: base.MetricsServiceName(dep), Namespace: d.Namespace}
```
Can this code churn be avoided?

pohly replied:
Yes, this can be avoided. There's no reason for moving these helper functions into the base package.

Partial patch:
```
diff --git a/pkg/apis/pmemcsi/base/base_types.go b/pkg/apis/pmemcsi/base/base_types.go
index 770600bcb..7bd6b2c11 100644
--- a/pkg/apis/pmemcsi/base/base_types.go
+++ b/pkg/apis/pmemcsi/base/base_types.go
@@ -349,12 +349,6 @@ func ControllerServiceName(d metav1.Object) string {
 	return GetHyphenedName(d) + "-controller"
 }
 
-// MetricsServiceName returns the name of the controller metrics
-// Service object used by the deployment
-func MetricsServiceName(d metav1.Object) string {
-	return GetHyphenedName(d) + "-metrics"
-}
-
 // ServiceAccountName returns the name of the ServiceAccount
 // object used by the deployment
 func ServiceAccountName(d metav1.Object) string {
diff --git a/pkg/apis/pmemcsi/v1beta1/deployment_types.go b/pkg/apis/pmemcsi/v1beta1/deployment_types.go
index 8cf85adb3..68c8b6eb9 100644
--- a/pkg/apis/pmemcsi/v1beta1/deployment_types.go
+++ b/pkg/apis/pmemcsi/v1beta1/deployment_types.go
@@ -158,3 +158,10 @@ func (d *Deployment) GetOwnerReference() metav1.OwnerReference {
 func (d *Deployment) HaveCertificatesConfigured() (bool, error) {
 	return d.Spec.HaveCertificatesConfigured()
 }
+
+// MetricsServiceName returns the name of the controller metrics
+// Service object used by the deployment
+func (d *Deployment) MetricsServiceName() string {
+	// TODO: also move GetHyphenedName and all other functions
+	return base.GetHyphenedName(d) + "-metrics"
+}
diff --git a/pkg/pmem-csi-operator/controller/deployment/controller_driver.go b/pkg/pmem-csi-operator/controller/deployment/controller_driver.go
index 346a5e3dc..3f13403e2 100644
--- a/pkg/pmem-csi-operator/controller/deployment/controller_driver.go
+++ b/pkg/pmem-csi-operator/controller/deployment/controller_driver.go
@@ -497,7 +497,7 @@ func (d *PmemCSIDriver) redeployMetricsService(r *ReconcileDeployment) (apirunti
 		object: func() apiruntime.Object {
 			return &corev1.Service{
 				TypeMeta:   metav1.TypeMeta{Kind: "Service", APIVersion: "v1"},
-				ObjectMeta: d.getObjectMeta(base.MetricsServiceName(d), false),
+				ObjectMeta: d.getObjectMeta(d.MetricsServiceName(), false),
 			}
 		},
 		modify: func(o apiruntime.Object) error {
@@ -653,7 +653,7 @@ func (d *PmemCSIDriver) HandleEvent(meta metav1.Object, obj apiruntime.Object, r
 		klog.Infof("Redeploying service: %s", meta.GetName())
 		if meta.GetName() == base.ControllerServiceName(d) {
 			_, err = d.redeployControllerService(r)
-		} else if meta.GetName() == base.MetricsServiceName(d) {
+		} else if meta.GetName() == d.MetricsServiceName() {
 			_, err = d.redeployMetricsService(r)
 		}
 		if err != nil {
diff --git a/test/e2e/operator/deployment_api.go b/test/e2e/operator/deployment_api.go
index 297707f63..8b32bcac1 100644
--- a/test/e2e/operator/deployment_api.go
+++ b/test/e2e/operator/deployment_api.go
@@ -620,7 +620,7 @@ var _ = deploy.DescribeForSome("API", func(d *deploy.Deployment) bool {
 				},
 				"metrics service": func(dep *api.Deployment) apiruntime.Object {
 					return &corev1.Service{
-						ObjectMeta: metav1.ObjectMeta{Name: base.MetricsServiceName(dep), Namespace: d.Namespace},
+						ObjectMeta: metav1.ObjectMeta{Name: dep.MetricsServiceName(), Namespace: d.Namespace},
 					}
 				},
 				"provisioner role": func(dep *api.Deployment) apiruntime.Object {
@@ -722,7 +722,7 @@ var _ = deploy.DescribeForSome("API", func(d *deploy.Deployment) bool {
 				},
 				"metrics service": func(dep *api.Deployment) apiruntime.Object {
 					obj := &corev1.Service{}
-					key := runtime.ObjectKey{Name: base.MetricsServiceName(dep), Namespace: d.Namespace}
+					key := runtime.ObjectKey{Name: dep.MetricsServiceName(), Namespace: d.Namespace}
 					EventuallyWithOffset(1, func() error {
 						return client.Get(context.TODO(), key, obj)
 					}, "2m", "1s").ShouldNot(HaveOccurred(), "get metrics service set")
```

avalluri replied:
> Can this code churn be avoided?

My initial changes were as you proposed. But, I felt like code duplication which we could avoid. I mean all(`alpha`, `beta`, and future) versions need not define the same function. So I moved them to the API `base` package.

My definition of base package is the 'common' code that could be shared by all versions without duplicating the code.

pohly replied:
> I mean all(alpha, beta, and future) versions need not define the same function.

Why? It's only called for the API revision that the operator code uses, so the function is only needed there. Or am I missing something?

The cleanest solution would be to not define those helper functions in the API package at all. They are not part of the *API*, they just support code in the operator which works with the API object.

pohly commented on test/e2e/operator/deployment_api.go:
```diff
@@ -744,6 +776,84 @@ var _ = deploy.DescribeForSome("API", func(d *deploy.Deployment) bool {
+		It("with default values", func() {
+			if d.HasOLM {
+				Skip("Conversion webhook is not supported by OLM")
+			}
```
The conversion *webhook* may not be supported, but *conversion* is supported - it just works a bit differently.

In other words, don't skip this and the other tests below for OLM and instead just update the expectations where necessary.

avalluri replied:
Yes, agree. I will modify the test.

avalluri replied:
Done.

avalluri commented on pkg/deployments/load.go:
```diff
@@ -23,7 +24,7 @@ import (
 
 // LoadObjects reads all objects stored in a pmem-csi.yaml reference file.
-func LoadObjects(kubernetes version.Version, deviceMode api.DeviceMode) ([]unstructured.Unstructured, error) {
+func LoadObjects(kubernetes version.Version, deviceMode base.DeviceMode) ([]unstructured.Unstructured, error) {
```
> We can avoid this code churn by importing `DeviceMode` into the v1beta1 API. If we ever remove `v1alpha` and `base`, then the rest of the code doesn't need to change again

So you wanted to say that DeivceMode is 'v1beta' specific and going to be removed in v1?. My intention is that we ever need to drop the base package. The definitions from this package are shared by all future versions also.

pohly replied:
> So you wanted to say that DeivceMode is 'v1beta' specific and going to be removed in v1?

No. I'm just saying that `DeviceMode` is part of the `v1beta1` API and we avoid code churn by exposing it as definition in that package. I have no plans to remove anything at this point.

> The definitions from this package are shared by all future versions also.

Only as long as they use exactly the same definitions. If we add another device mode that isn't supported by older PMEM-CSI instances, then we can no longer just update the base package.

pohly replied:
> If we add another device mode that isn't supported by older PMEM-CSI instances, then we can no longer just update the base package.

Actually, we already have that situation now. "fake" device mode is allowed in the v1beta1 API, but not in the v1alpha1 API.

avalluri replied:
I pushed the latest revision of PR that removes the additional `base` package. and avoids these changes.

pohly commented on deploy/kustomize/operator-webhook/kustomization.yaml:
```diff
@@ -0,0 +1,33 @@
+    fieldpath: metadata.namespace
+
+configurations:
+- kustomizeconfig.yaml
```
Nit: missing line break

avalluri replied:
Fixed.

pohly commented on deploy/kustomize/operator-webhook/kustomizeconfig.yaml:
```diff
@@ -0,0 +1,4 @@
+varReference:
+- kind: Certificate
+  group: cert-manager.io
+  path: spec/dnsNames
```
Nit: missing line break

avalluri replied:
Fixed.

pohly commented on deploy/kustomize/patches/crd-conversion-webhook-patch.yaml:
```diff
@@ -0,0 +1,23 @@
+          name: pmem-csi-operator-service
+          namespace: default # should match with the operator namespace
+          port: 9443
+          path: /convert
```
Nit: missing line break

avalluri replied:
Fixed.

avalluri commented on pkg/apis/pmemcsi/v1beta1/deployment_types_test.go:
```diff
@@ -0,0 +1,210 @@
+
+	Context("API", func() {
+		//
+		// TODO: Add input-validation tests for the API spec fields
```
I feel like this TODO is outdated and no additional such test required. IIRC this to was in the context of validating device mode value., which was covered in a different place. I removed the TODO.

pohly commented on deploy/kustomize/crd/pmem-csi.intel.com_deployments.yaml:
```diff
@@ -0,0 +1,543 @@
+
+---
```
Redundant blank line and item separator?

avalluri replied:
In fact, this file should not be committed, I will remove this.

pohly commented on deploy/kustomize/operator-webhook/kustomizeconfig.yaml:
```diff
@@ -0,0 +1,5 @@
+- kind: Certificate
+  group: cert-manager.io
+  path: spec/dnsNames
+
```
nit: extra line at the end.

avalluri replied:
Fixed.

pohly commented on deploy/kustomize/operator-webhook/webhook-certificate.yaml:
```diff
@@ -0,0 +1,25 @@
+    kind: Issuer
+    name: self-signed-issuer
+---
+
```
nit: extra separator and line at the end

avalluri replied:
Fixed.

pohly commented on deploy/kustomize/operator-webhook/webhook-service.yaml:
```diff
@@ -0,0 +1,12 @@
+      targetPort: 9443
+  selector:
+    name: pmem-csi-operator
+
```
nit: extra line

avalluri replied:
Fixed.

pohly commented on deploy/kustomize/patches/crd-conversion-webhook-patch.yaml:
```diff
@@ -0,0 +1,24 @@
+          namespace: default # should match with the operator namespace
+          port: 9443
+          path: /convert
+
```
nit: extra line

avalluri replied:
Fixed.

pohly commented on pkg/pmem-csi-operator/main.go:
```diff
@@ -16,13 +16,15 @@ import (
 
 	pmemcommon "github.com/intel/pmem-csi/pkg/pmem-common"
-
 	"github.com/operator-framework/operator-lib/leader"
```
nit: I find it useful to separate "our" imports from Go runtime imports (at the top) and third-party imports (below).

In other words, don't delete this blank line.

pohly replied:
Or rather, delete line 22 above this.

avalluri replied:
Fixed.

pohly reviewed (changes requested):
Mostly just some nits left now. Looks much better than with the "base" package (shorter diff!) and clean separation between different API versions.

pohly commented:
"conversion with explicit values" in "olm.API" is broken (https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/blue/organizations/jenkins/pmem-csi/detail/PR-826/15/tests).

avalluri commented on pkg/pmem-csi-operator/controller/deployment/deployment_controller.go:
```diff
@@ -14,7 +14,7 @@ import (
 	"time"
 
-	pmemcsiv1alpha1 "github.com/intel/pmem-csi/pkg/apis/pmemcsi/v1alpha1"
+	pmemcsiv1beta1 "github.com/intel/pmem-csi/pkg/apis/pmemcsi/v1beta1"
```
Done.

avalluri commented on pkg/pmem-csi-operator/controller/deployment/deployment_controller_test.go:
```diff
@@ -56,7 +56,7 @@ func getDeployment(d *pmemDeployment) *api.Deployment {
 		TypeMeta: metav1.TypeMeta{
 			Kind:       "Deployment",
-			APIVersion: "pmem-csi.intel.com/v1alpha1",
+			APIVersion: "pmem-csi.intel.com/v1beta1",
```
We have such a variable defined already named `SchemeGroupVersion` in each API version. I used that here and we need not touch in the future.

pohly commented:
Not sure what changed, but now "fedora-1_17.olm.API" is very unhappy (92 test failures in https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/blue/organizations/jenkins/pmem-csi/detail/PR-826/18/tests).

avalluri commented:
> Not sure what changed, but now "fedora-1_17.olm.API" is very unhappy (92 test failures in https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/blue/organizations/jenkins/pmem-csi/detail/PR-826/18/tests).

The error seems not related to this particular change. Somehow installed OLM fails to start the operator: 
```
+ /mnt/workspace/pmem-csi_PR-826/_work/bin/operator-sdk run packagemanifests --version 0.8.0 --namespace default --install-mode OwnNamespace --timeout 3m /tmp/tmp.W3oTimKXBK/olm-catalog
time="2020-12-09T12:40:39Z" level=info msg="Creating pmem-csi-operator registry"
time="2020-12-09T12:40:39Z" level=info msg=" Creating ConfigMap \"default/pmem-csi-operator-registry-manifests-package\""
time="2020-12-09T12:40:39Z" level=info msg=" Creating ConfigMap \"default/pmem-csi-operator-registry-manifests-0-8-0\""
time="2020-12-09T12:40:39Z" level=info msg=" Creating Deployment \"default/pmem-csi-operator-registry-server\""
time="2020-12-09T12:40:39Z" level=info msg=" Creating Service \"default/pmem-csi-operator-registry-server\""
time="2020-12-09T12:40:39Z" level=info msg="Waiting for Deployment \"default/pmem-csi-operator-registry-server\" rollout to complete"
time="2020-12-09T12:40:39Z" level=info msg=" Waiting for Deployment \"default/pmem-csi-operator-registry-server\" to rollout: 0 out of 1 new replicas have been updated"
time="2020-12-09T12:40:40Z" level=info msg=" Waiting for Deployment \"default/pmem-csi-operator-registry-server\" to rollout: 0 of 1 updated replicas are available"
time="2020-12-09T12:40:49Z" level=info msg=" Deployment \"default/pmem-csi-operator-registry-server\" successfully rolled out"
time="2020-12-09T12:40:49Z" level=info msg="Created CatalogSource: pmem-csi-operator-catalog"
time="2020-12-09T12:40:49Z" level=info msg="OperatorGroup \"operator-sdk-og\" created"
time="2020-12-09T12:40:49Z" level=info msg="Created Subscription: pmem-csi-operator-v0-8-0-sub"
time="2020-12-09T12:40:52Z" level=fatal msg="Failed to run packagemanifests: error approving install plan: Operation cannot be fulfilled on installplans.operators.coreos.com \"install-m4tpt\": the object has been modified; please apply your changes to the latest version and try again\n"
```

pohly commented:
> The error seems not related to this particular change. Somehow installed OLM fails to start the operator:

Can you file an upstream bug for this?

pohly commented on test/e2e/operator/deployment_api.go:
```diff
@@ -744,6 +769,84 @@ var _ = deploy.DescribeForSome("API", func(d *deploy.Deployment) bool {
+
+			validateDriver(deployment, true)
+
+			// Expect to ruturn alpha object converting from stored beta CR
```
I missed this when we merged the PR, but now I am looking at a test failure in this test after fixing the default log level and have some more questions.

Why does it say here "from stored beta CR"? The stored CR is in the v1alpha1 format, isn't it?

avalluri replied:
Stored version is always the latest(`spec.storage: true` in the CRD) version. And when the conversion happens to client requested version from that.

avalluri commented:
> > The error seems not related to this particular change. Somehow installed OLM fails to start the operator:
> 
> Can you file an upstream bug for this?

I filed the [bug](https://github.com/operator-framework/operator-sdk/issues/4308) and also submitted the [fix](https://github.com/operator-framework/operator-sdk/pull/4309).
